new presets system

ael-code opened the pull request:
I rewrote the presets parsing system. The code is a bit tricky but we now use a declarative code to teach the parser/checker, this means that adding new type field required no more then 4 static lines. Except the way we do parsing, almost all behavior about presets stay the same.
I wrote a lot of tests about skeleton parsing and prest_paths loading.
As we decided, I also moved presets module outside webant folder.

##### Notes:
 - For cli command to debug presets stuff refer to #99.

##### Evident change for librarian:
From the users point of view there are few changes:
 - in order to support different kind of input, has been introduced a new field ```type``` for property objects. At the moment we support two type : ```string``` and ```list```
 - properties are now an ordered list, so the librarian can decide how to organize form inputs in add-page directly from the json schema.
 - Exceptions and logs about preset stuff are now more informative ( you should understand exactly where did you miss ) 


##### Testing:
If you want to test preset you can use old method. You can use ```PRESET_PATHS``` conf to pass a list of path containing presets file. You can use example schemas from ```libreant/presets/defaultPresets/```


This should fix #70 and help with #59

boyska commented:
The `setup.py` seems to be wrong. It shows `webant.presets` module, which does not exist anymore

ael-code commented:
> The setup.py seems to be wrong. It shows webant.presets module, which does not exist anymore. How did you manage to python setup.py install it, btw? I could not.

Fixed before your comment ;)

boyska commented on presets/presetManager.py:
```diff
@@ -0,0 +1,223 @@
+            return
+
+        for i in range(len(paths)):
+            path = paths[i]
```
you mean `for path in paths`, right? :)

ael-code replied:
> you mean for path in paths, right? :)

We can change the cycle but the point is that I do not want to do side-effect on ```paths```. Will it  change something on performance side?

boyska replied:
Which side effect are you talking of?
It's not a matter of performance, which would be irrelevant. It's a matter of readability, which is way more important!

If you think that there is a good reason to write code like this, you should at least explain **why** in a comment. Any pythonista will look at this code wondering if it is just badly written, or if there is some specific reason it's written like this, which they are not being able to understand.
I still have not, for example :)

Other similar readability issues are commented separately

ael-code replied:
```for i in range(len(paths)):```
are you serious?

Anyway I will change it.

boyska replied:
serious about what? :)

Iterating on a list is 99.9999% of times done using the iteration builtin. Recurring to other method is commonly done when you have the need to _change_ the object you are iterating upon, or other not-very-clean usage of iterations.

Since there is a very procedural (in contrast to functional) style in the code, it's been hard for me to follow the code. I expected that inside some other method you were changing the paths, or stuff like that.

Now I guess this is not the case, so please write idiomatically!

ael-code replied:
Changed.

Anyway os.path.* modifies the path

boyska commented:
I think there are some "stylistic" problem that lead to decreased readability: PEP8 mandates `underscore_separated_method_names`, not `mixedCase`.

Also, as I pointed out in #99, some function should be pulled out of presetmanager and become functions.

I see the attempt to give meaningful errors to the user, but I suspect that this is exactly why it's so hard to make it modular. I don't have a solution on this, still I see some problem.

ael-code commented:
> I think there are some "stylistic" problem that lead to decreased readability: PEP8 mandates underscore_separated_method_names, not mixedCase.
> 

Please provide some specific example.

> Also, as I pointed out in #99, some function should be pulled out of presetmanager and become functions.
> 

This code is ready to support #99 implementation. See testing function for an example.

> I see the attempt to give meaningful errors to the user, but I suspect that this is exactly why it's so hard to make it modular. I don't have a solution on this, still I see some problem.

Solution to what? I did not understand the problem.

boyska commented on presets/presetManager.py:
```diff
@@ -0,0 +1,223 @@
+
+            # avoid empty string
+            if not path:
+                continue
```
is this important? why do we care? what if, for example, path is a list? why are we considering the case of empty string but not of non-strings?

ael-code replied:
we consider the case of non-strings in the following lines: os.path.* commands sequence raise error if input parameter is not of type string.

Empty strings instead will be transformed in ```.```

boyska replied:
ok, that's a good explanation. Please comment it!

boyska commented on presets/presetManager.py:
```diff
@@ -0,0 +1,223 @@
+            if depth == 0:
+                path = os.path.expanduser(path)  # replace ~
+                path = os.path.expandvars(path)  # replace vars
+                path = os.path.normpath(path)  # replace /../ and so on
```
why does this happen only if `depth == 0` ?

ael-code replied:
because if you apply those functions on successive iterations they simply will do nothing.

boyska replied:
I don't think that this is true. `normpath` will resolve symbolic links, so if you have a symbolic link inside the "base" directory, normpath will change it. Anyway, this doesn't make so much of a difference, and you could want to leave it this way. But it should at least be commented, clarifying to the reader why is there  such an exception, the fact that it is just an optimization and that there is nothing special about `depth == 0`

boyska commented on presets/presetManager.py:
```diff
@@ -0,0 +1,272 @@
+            return
+
+        for p in paths:
+            path = p
```
ok, this is getting weird :)
I suspected that the strange, C-inspired, style you were using before was just by change.
Now you're doing something similar again.

Can you explain why are you doing something like that? I guess there's a reason, and I cannot understand it.
Oh, btw, you might want to test this yourself

```python
l = [[x] for x in range(10)]

print l

for subl in l:
    asd = subl
    asd.append('ciao')

print l
```

ael-code replied:
@boyska you really need to be more specific!

You must keep in mind that I wrote this code, it makes sense for me and I did also tests.

If you find something strange, you have to explain me which part and why it is strange for you. 

> Can you explain why are you doing something like that? I guess there's a reason, and I cannot understand it.

In the previous comment there is not explicit subject and it's very difficult for me to guess what are you talking about. I've tried but I really can't.

boyska replied:
Ok, sorry. I wonder why you do ```for p in paths: path = p```
then you never use the `p` variable anymore. Why don't you simply do ```for path in paths```?
Is this some technique that I do not know?

I am vaguely convinced that you are trying to copy the path object to avoid modifying it. If I am correct, I encourage you to test the dummy example above, where something similar happens.

ael-code replied:
For strings seems to be different:

```python
import os

l = ["","/tmp/../tmp/","/home///ael/./././"]
print l
for p in l:
    path = p
    path = os.path.normpath(path)
print l

for p in l:
    p = os.path.normpath(p)
print l
```
Both for cycles do not modify objects in ```l```.

boyska replied:
>For strings seems to be different:

Strings are immutable, you can't expect them to change in any case.

>    path = os.path.normpath(path)

Indeed, you are not even trying to change them, you are just making the
variable point to something else

I think that you should better understand how variables works in Python.
They're actually more similar to names than to "what C family calls variables"

When you do
```
a = "foo"
b = a
a = "foozah"
```
what are you expecting to see?

>Both for cycles do not modify objects in ```l```.

...But we're losing the focus. I was wondering why you wrote such code.
I proposed the hypothesis that you want to avoid side-effects, and
explained you that even in that case, the technique you used was giving
no guarantees in that respect. Now you're pointing out yet another
reason why there are no side effects. Ok. But then why is the code
written like that?  Every time I read it, I feel dumb :(

A more concrete proposal:
```python
normpaths = (os.path.expanduser(os.path.expandvars(os.path.normpath(p))) for p in paths if p)

for path in normpaths:
    #all the rest of the code
```

I think this is shorter, clearly has no side-effect, separate the normalization part from the loading part. There are probably many other ways to write the same thing, find the one that you like more.

ael-code replied:
I changed the code. 
I have not followed your suggestion due to readability, anti-cvs and bad performance (double cycle). Yes I know that performance in this case does not matter but I prefer to write fast code anyway.

ael-code commented:
All the issues raised until now, have been fixed/discussed.
Do you confirm?

boyska commented:
Nope :) but they disappeared from the github thing.

I expressed concerns over difficult to understand code. You explained the rationale for most of your choices, but they were (for me, at least) non-trivial and contributed to confusion.
Explaining them in comments would make code clearer. I am here referring to the check for path not to be empty, and for the check about `depth == 0`. About that check: I suppose it is wrong, but I still have to produce a failing example.

ael-code commented:
> but they disappeared from the github thing.

what?

> About that check: I suppose it is wrong, but I still have to produce a failing example.

I dare you ;)

boyska commented:
Ok, here it is:
```
mkdir -p /tmp/presets/
ln -s $(realpath presets/defaultPresets/bookSkeleton.json) /tmp/presets/
WEBANT_PRESETS_PATH=/tmp/presets/ libreant-manage shell
```

inside the shell, run

`app.presetManager.presets` and you will find nothing. The problem is that using that `depth == 0` will prevent symbolic links to be followed

ael-code commented:
ahah, maybe you wanna try with ```WEBANT_PRESET_PATHS``` instead of ```WEBANT_PRESETS_PATH```.

I double dare you :)

boyska commented:
lol :) well, now I tried with the correct option and, just to be sure, also made a copy.
```sh
sp presets/defaultPresets/bookSkeleton.json /tmp/presets/copy.json
```

the `copy` is parsed, but the symlink is not.

ael-code commented:
> the copy is parsed, but the symlink is not.

How do you know that symlink is not parsed?
In log messages you should find something like this:
```
[presets.presetManager] [ERROR] Skipping preset path: "some_path" [ Duplicate skeleton id: some_id ]
```

boyska commented:
Oh, the other problem I raised is the lack of **definition** of things. What's a Preset? What's a Property? what's a skeleton? who cares about what? What's the behaviour of `_loadPreset` exactly? why does the warning raised by loadPreset talk about "skeletons" ? In the case of duplicate skeleton id, who gets the prececedence? is it deterministic, or not (not being is a valid solution, but should be made explicit)?

Another issue is the PEP8 consistency: some method follow the recommended `under_score_form`, like for exemple `Preset.check_id`. Many other follow `mixedCase`, like `Preset.getProperty`, or `PresetManager._loadPaths`.

Another part of code that is really intricated is the `_warn`. With `_warn` you immediately decide, based on `self.strict` if you should raise an exception or just log an error. I can't see what's the advantage of doing this immediately instead of just raising an exception and handling it in the calling code. This also leads to strange control flow, let me explain: when strict is True, a `_warn`  will produce an exception, so the function will exit. When strict is False, there is no exception, so there is no exit. This is why you make almost every `_warn` followed by a return. I said "almost". Some of those return are not there, but they are the last line in the control flow, so it's dirty but has no semantic difference. Some are followed by a `continue`, despite the same behaviour would have been accomplished without doing so.
I think that just raising exceptions immediately would be much clearer. It's up to `_loadPaths` to handle exceptions and decide if it should just be a warning or not.

Another thing that I do not like is the recursion in `_loadPaths`: wouldn't it be simpler to first generate the list of files, and then do `self.presets = map(self._loadPreset, allfiles)` ?

boyska commented:
why are you expecting that it has a duplicate id? it's not even parsed, because it's a symlink, and you're explicitly refusing symlinks, except if they're inside the root directory.

ael-code commented:
> it's not even parsed, because it's a symlink, and you're explicitly refusing symlinks, except if they're inside the root directory.

I did some tests and symlinks seem to work.
please, can you provide steps to reproduce the problem you are talking about?

boyska commented:
Ok.
```sh
mkdir -p /tmp/presets/subdir/asd/
 ln -s presets/defaultPresets/bookSkeleton.json /tmp/presets/subdir/asd
WEBANT_PRESET_PATHS=/tmp/presets/ libreant-manage shell 
```

inside the shell, run `app.presetManager.presets` to see the result. I only see `{}`
If you want to be sure that the directory is being parsed, do

```
cp presets/defaultPresets/bookSkeleton.json /tmp/presets/subdir/copied.json
vim /tmp/presets/subdir/copied.json
```

and change the id of the `copied.json` (otherwise it could conflict)

ael-code commented:
> The problem is that using that depth == 0 will prevent symbolic links to be followed

> why are you expecting that it has a duplicate id? it's not even parsed, because it's a symlink, and you're explicitly refusing symlinks, except if they're inside the root directory.

The above sentences are wrong. The behaviour you are talking about actually happens but it is not a bug and **IT'S ABSOLUTELY NOT RELATED TO SYMLINKS**. All files/folders under certain depth level are ignored, this is controlled by ```PresetManager.MAX_DEPTH``` constant. The constant was set at level 2, maybe this is too low, I am now upgrading this value to 5. ( who are going to organize presets in a 4 depth folder tree? )
We can now stop fantasy story about symlinks ;)

boyska commented:
Good point. So I increased MAX_DEPTH to 5, and also I added a symlink inside `subdir/` instead of `subdir/asd/`. I can confirm that it still happens.

boyska commented:
By the way, what's the rationale about setting a max_depth?

boyska commented:
(sorry for the comment flood). I have now seen that you rebased, changing the `MAX_DEPTH`, so I re-tested. Still no solution to the bug. By the way, the code is pretty clear, so I can't see why you are not just patching it. The code says that if it is not a directory not a file, it's `not a regular file`. A symlink is not a regular file, so how can you expect it to work?

boyska commented:
Ok, I found the bug. It was about the target specification. If you do
```sh
ln -s $(realpath presets/defaultPresets/bookSkeleton.json) /tmp/presets/subdir/asd/real.json
```
it will work. Relative symlinks are odd!

I also found my own bug when reading the code: I assumed that `isfile` does not follow symlink, while it does. So ok, the symlink issue is solved.

ael-code commented:
symlinks fantasy story end. :)

> Another part of code that is really intricated is the _warn. With _warn you immediately decide, based on self.strict if you should raise an exception or just log an error. I can't see what's the advantage of doing this immediately instead of just raising an exception and handling it in the calling code. This also leads to strange control flow, let me explain: when strict is True, a _warn will produce an exception, so the function will exit. When strict is False, there is no exception, so there is no exit. This is why you make almost every _warn followed by a return. I said "almost". Some of those return are not there, but they are the last line in the control flow, so it's dirty but has no semantic difference. Some are followed by a continue, despite the same behaviour would have been accomplished without doing so.
> I think that just raising exceptions immediately would be much clearer. It's up to _loadPaths to handle exceptions and decide if it should just be a warning or not.

You're right, this part is a bit tricky.
As soon as I'll have some free time, I'll try to reorganize the logging/exceptions stuff

boyska commented:
On Tue, Feb 24, 2015 at 04:14:47AM -0800, ael wrote:
>symlinks fantasy story end. :)

yep, sorry for making you waste time. For me it was not wasted, since I
understood the code better ;)

ael-code commented:
> yep, sorry for making you waste time

No problem

> Relative symlinks are odd!

You can use ```ln``` to create relative symbolic links with ```-r, --relative``` option. I personally do not use that.

ael-code commented:
> Another part of code that is really intricated is the _warn. With _warn you immediately decide, based on self.strict if you should raise an exception or just log an error. I can't see what's the advantage of doing this immediately instead of just raising an exception and handling it in the calling code. This also leads to strange control flow, let me explain: when strict is True, a _warn will produce an exception, so the function will exit. When strict is False, there is no exception, so there is no exit. This is why you make almost every _warn followed by a return. I said "almost". Some of those return are not there, but they are the last line in the control flow, so it's dirty but has no semantic difference. Some are followed by a continue, despite the same behaviour would have been accomplished without doing so.
> I think that just raising exceptions immediately would be much clearer. It's up to _loadPaths to handle exceptions and decide if it should just be a warning or not.

Oke, I looked at the code again; it is very difficult to change it:
In the default way, ```strict``` is false and all ```_warn()``` will be turned in a log message, I use log messages because after the call I can continue the function flow ( this is the reason for the ```continue``` statement). If I substitute log calls with exceptions as you suggested it will be difficult to continue with the function flow. Do you have some idea ?

Anyway if you look at ```_warn()``` like a log call the code became pretty standard.

boyska commented:
> I use log messages because after the call I can continue the function flow 

If I am not wrong, every time you call `self._warn` you also call `return`. You continue the program flow in _the caller function_. And this can be done using

```python
try:
    self._loadPath(fname)
except WhatException as exc:
    logger.warning('WAT!')
    if self.strict:
        raise exc
```

ael-code commented:
new changes:

- reorganized exceptions handling
- deleted all reference to ```skeleton``` word in the code

boyska commented:
Seems a big improvement. One minor thing: I don't like very much using `except ExceptionClass, e`, and prefer `except ExceptionClass as e` instead. I'm not alone: http://stackoverflow.com/a/2535770 , https://www.python.org/dev/peps/pep-3110/

Now I noticed that there is no way to specify the strict mode. A `PRESET_STRICT` variable would be easy to plug in.

One thing that I don't like is that you call `_warn`, which in turn can raise an exception if `strict is True`. The problem about this is that, while the message is informative, the stacktrace becomes completely useless. This is ok for a user, but not for a developer.
To test what I mean, create a file called `broken.json` with `{` as content, and set strict to True manually in the code.
You'll get something like this

```
Traceback (most recent call last):
  File "/home/gordo/.virtualenvs/libreant-test/bin/libreant-manage", line 9, in <module>
    load_entry_point('libreant==0.1', 'console_scripts', 'libreant-manage')()
  File "/home/gordo/.virtualenvs/libreant-test/lib/python2.7/site-packages/pkg_resources.py", line 356, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/gordo/.virtualenvs/libreant-test/lib/python2.7/site-packages/pkg_resources.py", line 2431, in load_entry_point
    return ep.load()
  File "/home/gordo/.virtualenvs/libreant-test/lib/python2.7/site-packages/pkg_resources.py", line 2147, in load
    ['__name__'])
  File "/tmp/pr.libreant-100-pL7_jU/webant/manage.py", line 16, in <module>
    app = LibreantCoreApp('webant', conf)
  File "/tmp/pr.libreant-100-pL7_jU/webant/webant.py", line 41, in __init__
    self.presetManager = PresetManager(self.config['PRESET_PATHS'])
  File "/tmp/pr.libreant-100-pL7_jU/presets/presetManager.py", line 36, in __init__
    self._loadPaths(paths)
  File "/tmp/pr.libreant-100-pL7_jU/presets/presetManager.py", line 67, in _loadPaths
    self._loadPaths([os.path.join(path, child)], depth + 1)
  File "/tmp/pr.libreant-100-pL7_jU/presets/presetManager.py", line 67, in _loadPaths
    self._loadPaths([os.path.join(path, child)], depth + 1)
  File "/tmp/pr.libreant-100-pL7_jU/presets/presetManager.py", line 75, in _loadPaths
    self._warn(path, str(e))
  File "/tmp/pr.libreant-100-pL7_jU/presets/presetManager.py", line 106, in _warn
    raise PresetException(errMex)
presets.presetManager.PresetException: Failed to load preset: "/tmp/presets/sublink/broken.json" [ JSON decoding error: Expecting object: line 1 column 2 (char 1) ]
```

can you see the point? `self._loadPath` is not even mentioned!
I propose, as an alternative, to just re-raise the same exception.

ael-code commented:
> Seems a big improvement. One minor thing: I don't like very much using except ExceptionClass, e, and prefer except ExceptionClass as e instead. I'm not alone: http://stackoverflow.com/a/2535770 , https://www.python.org/dev/peps/pep-3110/

Got it.

> Now I noticed that there is no way to specify the strict mode. A PRESET_STRICT variable would be easy to plug in.

I do not like to provide this possibility. I think implementing #99 will be better.

> The problem about this is that, while the message is informative, the stacktrace becomes completely useless.

I know that. Fixing...

boyska commented:
Mh, a command to check presets is surely useful, but it won't make your application run in "strict mode"; that is, exploding if some file is not ok. If the "strict" thing is only meant to appear on a separate tool, then... how strict is it? I can't understand the usecase, now.

ael-code commented:
> Mh, a command to check presets is surely useful, but it won't make your application run in "strict mode"; that is, exploding if some file is not ok. If the "strict" thing is only meant to appear on a separate tool, then... how strict is it? I can't understand the usecase, now.


can you provide some real case in which you will set strict from config instead of using cli command ?

boyska commented:
lol ok this is getting strange.
You implemented a presetmanager, with a `strict` option. I don't think this option is much important, but you put a lot of effort into it, and lot of code changes a bit because of this "double path", so ok. Now you say that noone will use the `strict` option, except for manual checking.

But why should manual checking need strictness? Aren't error messages enough?

ael-code commented:
I cleaned up the code a lot. Now it is very simple. I hope you like the new solution.

> Seems a big improvement. One minor thing: I don't like very much using except ExceptionClass, e, and prefer except ExceptionClass as e instead. I'm not alone: http://stackoverflow.com/a/2535770 , https://www.python.org/dev/peps/pep-3110/

fixed

> The problem about this is that, while the message is informative, the stacktrace becomes completely useless.

fixed

ael-code commented:
strict parameter was primary introduced because of testing ( testing function use ```strict=True```). I don't know how to write test that catch log messages. 
Strict parameter it will be also useful for cli preset-check implementation. 
I don't think we need also a config entry.

boyska commented:
Ok, that's a good explanation. But then you are testing a different program flow than the actual one :)
Which is why I was proposing a `validate_path` top-level function that will _return_ a list of errors. Then you test mainly that function, and use it inside presetManager.

Anyway, the new code is much simpler, but I am not truly convinced that it works properly, so I will check.

boyska commented:
These things are missing:
* adhere to PEP8: method names should me `lower_case_joined_with_underscore`
* remove `getProperty` method: not useful, not documented
* improve unit testing:
  * `check_id` never fails
  * no I/O errors
  * no json validation errors
  * more generally, follow what nose-coverage shows you!
* improve documentation
  * a glossary would be very useful; especially, the role of each class should be clearer
  * clarify what the objects _won't do_: especially, clarify that there is no data validation (still)

ael-code commented:
last changes:
 - all function names follow PEP8 style
 - ```getProperty``` removed
 - added ```Preset.validate()``` function
 - improoved unit testing ( I/O Error not tested, I do not know how to simulate that )
 - improoved documentation

boyska commented:
I discovered that adding two properties with the same id is not an error at all. Is it normal?

ael-code commented:
> I discovered that adding two properties with the same id is not an error at all. Is it normal?

nope. This is a bug.
I will fix this with an exception/warning during preset schema validation and I will skip all the duplicate except the first one on the array.

boyska commented:
please consider merging https://github.com/ael-code/libreant/pull/1 before fixing it. I can foresee merging hell otherwise. I'm pretty sure you will agree with what I am proposing in that pull request

ael-code commented:
```test_properties_type_enum_values_nonstring``` tests almost same code of ```test_properties_type_enum_values_type```

boyska commented:
This is not true; it follows a different path.
On `values_type`, the exception is raised inside `Schema._verify_fields`, while on `values_nonstring` it is raised inside `Property.check_values`. The exception message is also different.

I have now changed `values_type` to highlight the difference with `values_nonstring`.

boyska commented:
yesterday we agreed that testing non essential things is not important, and should not be done.

ael-code commented:
> I discovered that adding two properties with the same id is not an error at all. Is it normal?

solved and added relative test.
 If there are properties with the same id in the same preset a ```PresetException``` is raised.

boyska commented:
When you scan for files, you accept every file, despite of extension, filename, etc. I don't like this very much, because `.*.swp` files are also considered. One solution could be to just consider `*.json` files. or to ignore files starting with `.` (but this is very unix-specific!)

boyska commented:
using `bookSkeleton.json` preset, upload is not allowed, despite the file content says the opposite. I have not done any further debugging.

ael-code commented:
> When you scan for files, you accept every file, despite of extension, filename, etc. I don't like this very much, because .*.swp files are also considered. One solution could be to just consider *.json files. or to ignore files starting with . (but this is very unix-specific!)

solved. Now only files with ".json" suffix are considered.

> using bookSkeleton.json preset, upload is not allowed, despite the file content says the opposite. I have not done any further debugging.

fixed. The template was looking for ```allowUpload``` instead of ```allow_upload```

boyska commented:
the `Preset.validate()` is not used in webant, is it? I can bypass any validation just doing like this:
```sh
 curl -XPOST 'localhost:5000/add?preset=book' -d 'ctg=novel' -d '_language=it'
```

while the "book" presets should require the title to be present.

boyska commented:
apart from the above comment (which solution you might decide to postpone in a separate pull request), I have no objections to merging.

ael-code commented:
On Mar 9, 2015 10:16 AM, "BoySka" <notifications@github.com> wrote:
>
> the Preset.validate() is not used in webant, is it? I can bypass any
validation just doing like this:
>
>  curl -XPOST 'localhost:5000/add?preset=book' -d 'ctg=novel' -d
'_language=it'
>
> while the "book" presets should require the title to be present.
>
Preset.validate() is not used at the moment. I will implement it in webant
in another pull request. ( maybe an issue should be opened as remainder )

boyska commented:
cheers
